Event scanner robustness

derekpierre opened the pull request:
**Type of PR:**
- [x] Bugfix
- [ ] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [x] 2
- [ ] 3

**What this does:**
Closes https://github.com/nucypher/sprints/issues/243.

Related to issues we've noticed with Alchemy free tier.

_(From Copilot)_
- Adds retry logic with configurable chunk size reduction for failed RPC requests
- Implements special detection and handling for Alchemy free tier limitations
- Introduces environment variable configuration for chunk size limits

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**

~~Don't merge PR until `tapir` testing is completed. See https://github.com/nucypher/nucypher/pull/3647#issuecomment-3307643722~~

derekpierre committed:
Set max chunk size to 1,000 instead of 10,000.

derekpierre committed:
Handle issues during event scanning better with block range adjustments, retries and delays.
The EventScanner has to be adaptable since rpc endpoints can experience issues retrieving events due to rate/size limit/block range restrictions.

derekpierre committed:
Re-use chunk size that successfully worked after decrease.

derekpierre committed:
Add additional check and adjustment when encountering alchemy free tier event logs limits.
Add a few env variables to provide us with some additional flexibility.

Co-authored-by: Manuel Montenegro <manuel@nucypher.com>
Co-authored-by: KPrasch <kieranprasch@gmail.com>

derekpierre committed:
Add enforcement of min/max scan chunk size for EventScanner.
Updated unit tests accordingly.

derekpierre committed:
Add tests for _get_logs processing including http errors, retries and alchemy free tier checks.

manumonti committed:
Apply ruff formatting

KPrasch commented on nucypher/config/constants.py:
```diff
@@ -17,6 +19,13 @@
+NUCYPHER_ENVVAR_ALCHEMY_FREE_TIER_MAX_CHUNK_NUM_BLOCKS = (
+    "NUCYPHER_ALCHEMY_FREE_TIER_MAX_CHUNK_NUM_BLOCKS"
+)
+NUCYPHER_ENVVAR_MIN_CHUNK_NUM_BLOCKS = "NUCYPHER_MIN_CHUNK_NUM_BLOCKS"
+NUCYPHER_ENVVAR_MAX_CHUNK_NUM_BLOCKS = "NUCYPHER_MAX_CHUNK_NUM_BLOCKS"
```
I suppose env vars are competing for ursula config files / CLI launch flags.  Not opposed to using both,  but figured I'd mention it in case there are any more broad implications for diverging these two sets of settings.

derekpierre replied:
I can definitely see that intuition.

My concern is around having a setting that persists in the configuration vs being more temporary via an env variable. A launch flag could be interesting, but feels very public since it'll be in the help output.

It's tricky - this isn't so much an ongoing user-facing configuration, but more of a temporary and precautionary measure in case there are any issues with the default values we've chosen. It allows us to address any problems immediately without the need for a release to fix, but the eventual release would probably not necessitate setting the environment variable anymore.

KPrasch commented on nucypher/utilities/events.py:
```diff
@@ -71,6 +79,36 @@ def write_events_to_csv_file(csv_file: Path,
+        # very specific error case
+        rpc_response = http_error.response.json()
+        is_http_400 = http_error.response.status_code == HTTPStatus.BAD_REQUEST
+        is_alchemy_provider = "alchemy" in getattr(web3.provider, "endpoint_uri", "")
```
When might web3 provider not have this attribute?  in tests?

derekpierre replied:
Not sure, but I noticed that we took this approach with middleware - https://github.com/nucypher/nucypher/blob/main/nucypher/blockchain/eth/clients.py#L93.

KPrasch commented on nucypher/utilities/events.py:
```diff
@@ -71,6 +79,36 @@ def write_events_to_csv_file(csv_file: Path,
+                is_rpc_bad_request,
+            ]
+        )
+    except (ValueError, KeyError):
```
This handles the JSONDecode (or whichever may arise from the .json() method)?

derekpierre replied:
Yep - `JSONDecodeError` subclasses `ValueError`.

KPrasch commented on nucypher/utilities/events.py:
```diff
@@ -404,27 +442,48 @@ def _get_logs(
-                # update to block since range could be problematic; don't go lower than 5 blocks
+
+            # Assumption: the reason for http error is fetching too many blocks
+            if i >= max_retries - 1:
```
renaming `i` to `attempt` will read nicely here.

derekpierre replied:
✔️

KPrasch commented on nucypher/utilities/events.py:
```diff
@@ -404,27 +442,48 @@ def _get_logs(
+                if i > 0:
+                    # we already reduced the chunk size for Alchemy, but it did not help
+                    logger.warn(
+                        "Alchemy free tier reduction was unsuccessful, retrying will not help"
```
What a corner case!

KPrasch commented on nucypher/utilities/events.py:
```diff
@@ -404,27 +442,48 @@ def _get_logs(
-            if i < max_retries - 1:
-                # update to block since range could be problematic; don't go lower than 5 blocks
+
+            # Assumption: the reason for http error is fetching too many blocks
```
Perhaps we check for a 400s series error after all?

derekpierre replied:
We do now for the alchemy free tier check - https://github.com/nucypher/nucypher/pull/3647/files#diff-b26c3379a039bbbd7265cebf12878a56915fa3a99899bcb4af7ce5d4eb526b89R99

manumonti committed:
Enable json error test

manumonti committed:
Add test for get logs when connection error

manumonti commented:
I added one more test, take a look @derekpierre 

The PR looks good enough to be reviewed and accepted, in my opinion.

derekpierre committed:
Rename `i` variable to `attempt` for clarity.

derekpierre commented:
> I added one more test, take a look @derekpierre

👍 

> The PR looks good enough to be reviewed and accepted, in my opinion.

A few things before merging:
- ~~Newsfragment~~
- ~~We need to look into an overall env variable for the min_chunk_size for the ActiveRitualTracker~~
- ~~Ideally improve test coverage - We are at 70% of the diff in coverage (< 85%)~~
- ~~We generate a docker build from this PR and put it on **`tapir` (NOT `lynx`)** for some sanity testing. One of the `tapir` nodes should be changed to use an Alchemy free tier endpoint.~~

derekpierre committed:
Add bugfix newsfragment for #3647.

derekpierre committed:
One environment variable to rule them all (for dkg ritual). Allow setting of min chunk size specifically for DKG rituals that basically allows overriding min chunk size for event scanning.

derekpierre committed:
Used named constants for default parameters used by EventScanner.
Update method documentation.

derekpierre committed:
Update event scanning tests to use higher level call to scan_chunk() instead of lower level _get_log() call; this improves test coverage.
Add test to ensure that alchemy free tier is only retried once regardless of number of retries, since retrying doesn't change the chunk size.

derekpierre committed:
Rename parameters and constants for clarity.

derekpierre committed:
Improve error messaging for chunk size decrease factor.
Add value check for chunk size increase factor.
Add testst.

derekpierre committed:
Add test ensuring that reduced chunk size is used for scanning the remaining chunks in the block range.

derekpierre committed:
Ensure that environment variables are casted to int.

derekpierre commented on nucypher/utilities/events.py:
```diff
@@ -30,37 +35,55 @@
+)  # they say 10, but it's really < 10
+# some reasonable minimum below alchemy free tier; we don't actually want this too low
+MIN_CHUNK_NUM_BLOCKS = os.environ.get(NUCYPHER_ENVVAR_MIN_CHUNK_NUM_BLOCKS, 9)
+MAX_CHUNK_NUM_BLOCKS = os.environ.get(NUCYPHER_ENVVAR_MAX_CHUNK_NUM_BLOCKS, 1000)
```
✔️  - Great catch Copilot.

derekpierre commented on nucypher/utilities/events.py:
```diff
@@ -360,9 +446,89 @@ def scan(
+            logs = web3.eth.get_logs(event_filter_params)
+            return logs, to_block_to_use
+        except HTTPError as http_error:
+            logger.warn(
+                f"eth_getLogs API call failed for range {from_block} - {to_block_to_use} ({to_block_to_use - from_block} blocks) on attempt {attempt + 1}/{max_retries}: {http_error}"
+            )
```
❌ - this is twisted's logger, not python's.

derekpierre commented on nucypher/utilities/events.py:
```diff
@@ -360,9 +446,89 @@ def scan(
+            # Assumption: the reason for http error is fetching too many blocks
+            if attempt >= max_retries - 1:
+                # no more retries left
+                logger.warn("eth_getLogs API call failed, no more retries left")
```
❌ - this is twisted's logger, not python's.

derekpierre commented:
> * We generate a docker build from this PR and put it on **`tapir` (NOT `lynx`)** for some sanity testing. One of the `tapir` nodes should be changed to use an Alchemy free tier endpoint.

^ @manumonti , `tapir` has been updated to run the latest code. Can you run some DKG tests on `tapir`? Basically something like:
- Completion of a couple short term DKG rituals
- Resulting rituals can be used for successful encryption/decryption

And any other tests you deem prudent.

manumonti commented:
@derekpierre 

I've run some tests over tapir network.

Nodes are running v7.6.1.

✅ Initiate Ritual with Alchemy Paid Subscription endpoint with all nodes running v7.6.1 (ritual 15)

```
2025-09-18T14:08:11+0000 [RitualTracker#debug] (0x91d12A) Scanning events in block range 26573400 - 26573480
2025-09-18T14:08:11+0000 [EventActuator#debug] Scanning for blocks: 26573400 - 26573460, chunk size 60, last chunk scan took 0, last logs found 0
2025-09-18T14:08:11+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26573400, 'toBlock': 26573460}
2025-09-18T14:08:11+0000 [EventActuator#debug] Scanning for blocks: 26573461 - 26573480, chunk size 120, last chunk scan took 0.07992291450500488, last logs found 0
2025-09-18T14:08:11+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26573461, 'toBlock': 26573480}
```

✅ Encrypt/decrypt using ritual 15 with Alchemy Paid Subscription endpoint

```
2025-09-18T19:22:54+0000 [condition-eval#info] Evaluating access conditions {'version': '1.0.0', 'condition': {'chain': 80002, 'method': 'eth_getBalance', 'parameters': ['0x210eeAC07542F815ebB6FD6689637D8cA2689392', 'latest'], 'returnValueTest': {'comparator': '==', 'value': 0}, 'conditionType': 'rpc'}}
```

✅ A node (Alchemy Paid subscription endpoint) is stopped for a while and then, when it starts again, the RitualTracker has to catch up with Alchemy Paid Subscription endpoint, so the event scanner has to get a large number of blocks. The  chunk size scale up in every call.

```
2025-09-18T19:42:57+0000 [RitualTracker#debug] (0x131617) Scanning events in block range 26581705 - 26583505
2025-09-18T19:42:57+0000 [EventActuator#debug] Scanning for blocks: 26581705 - 26581765, chunk size 60, last chunk scan took 0, last logs found 0
2025-09-18T19:42:57+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26581705, 'toBlock': 26581765}
2025-09-18T19:42:57+0000 [EventActuator#debug] Scanning for blocks: 26581766 - 26581886, chunk size 120, last chunk scan took 0.15644121170043945, last logs found 0
2025-09-18T19:42:57+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26581766, 'toBlock': 26581886}
2025-09-18T19:42:57+0000 [EventActuator#debug] Scanning for blocks: 26581887 - 26582127, chunk size 240, last chunk scan took 0.16739678382873535, last logs found 0
2025-09-18T19:42:57+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26581887, 'toBlock': 26582127}
2025-09-18T19:42:57+0000 [EventActuator#debug] Scanning for blocks: 26582128 - 26582608, chunk size 480, last chunk scan took 0.17091846466064453, last logs found 0
2025-09-18T19:42:57+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26582128, 'toBlock': 26582608}
2025-09-18T19:42:58+0000 [operator#debug] ritual #15 is not waiting for transcripts; status=5; skipping execution
2025-09-18T19:42:58+0000 [operator#debug] No action required for phase 1 of DKG protocol for some reason or another.
2025-09-18T19:42:58+0000 [EventActuator#debug] Processing event StartRitual, block: 26582339
2025-09-18T19:42:58+0000 [operator#debug] ritual #15 is not waiting for aggregations; status=5.
2025-09-18T19:42:58+0000 [EventActuator#debug] Processing event StartAggregationRound, block: 26582396
2025-09-18T19:42:58+0000 [RitualTracker#debug] Event 'EndRitual', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-18T19:42:58+0000 [EventActuator#debug] Processing event EndRitual, block: 26582457
2025-09-18T19:42:58+0000 [EventActuator#debug] Scanning for blocks: 26582609 - 26582669, chunk size 60, last chunk scan took 0.8159382343292236, last logs found 0
2025-09-18T19:42:58+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26582609, 'toBlock': 26582669}
2025-09-18T19:42:58+0000 [EventActuator#debug] Scanning for blocks: 26582670 - 26582790, chunk size 120, last chunk scan took 0.15748977661132812, last logs found 0
2025-09-18T19:42:58+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26582670, 'toBlock': 26582790}
2025-09-18T19:42:59+0000 [EventActuator#debug] Scanning for blocks: 26582791 - 26583031, chunk size 240, last chunk scan took 0.1859877109527588, last logs found 0
2025-09-18T19:42:59+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26582791, 'toBlock': 26583031}
2025-09-18T19:42:59+0000 [EventActuator#debug] Scanning for blocks: 26583032 - 26583505, chunk size 480, last chunk scan took 0.1637887954711914, last logs found 0
2025-09-18T19:42:59+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe']], 'fromBlock': 26583032, 'toBlock': 26583505}
2025-09-18T19:42:59+0000 [RitualTracker#debug] Scanned total of 3 events, in 1.89711332321167 seconds, total 8 chunk scans performed
```

✅ Three of five nodes are restarted using Alchemy free tier, so the ritual tracker has to catch up scanning a lot of blocks. Alchemy free tier detected on the three.

```
2025-09-18T20:01:48+0000 [RitualTracker#debug] (0x131617) Scanning events in block range 26582271 - 26584071
2025-09-18T20:01:48+0000 [EventActuator#debug] Scanning for blocks: 26582271 - 26582331, chunk size 60, last chunk scan took 0, last logs found 0
2025-09-18T20:01:48+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278']], 'fromBlock': 26582271, 'toBlock': 26582331}
2025-09-18T20:01:50+0000 [EventActuator#warn] eth_getLogs API call failed for range 26582271 - 26582331 (60 blocks) on attempt 1/3: 400 Client Error: Bad Request for url: https://polygon-amoy.g.alchemy.com/v2/<redacted>
2025-09-18T20:01:50+0000 [EventActuator#warn] Alchemy free tier detected. Retrying with range 26582271 - 26582280 (9 blocks)
2025-09-18T20:01:50+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278']], 'fromBlock': 26582271, 'toBlock': 26582280}
2025-09-18T20:01:50+0000 [EventActuator#debug] Scanning for blocks: 26582281 - 26582290, chunk size 9, last chunk scan took 1.4974768161773682, last logs found 0
```

✅ New ritual (16) with three of five nodes using Alchemy free tier endpoint

```
2025-09-18T20:10:53+0000 [EventActuator#debug] Scanning for blocks: 26584263 - 26584323, chunk size 60, last chunk scan took 0, last logs found 0
2025-09-18T20:10:53+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26584263, 'toBlock': 26584323}
2025-09-18T20:10:55+0000 [EventActuator#warn] eth_getLogs API call failed for range 26584263 - 26584323 (60 blocks) on attempt 1/3: 400 Client Error: Bad Request for url: https://polygon-amoy.g.alchemy.com/v2/<redacted>
2025-09-18T20:10:55+0000 [EventActuator#warn] Alchemy free tier detected. Retrying with range 26584263 - 26584272 (9 blocks)
2025-09-18T20:10:55+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26584263, 'toBlock': 26584272}
2025-09-18T20:10:55+0000 [EventActuator#debug] Scanning for blocks: 26584273 - 26584282, chunk size 9, last chunk scan took 1.4286384582519531, last logs found 0
2025-09-18T20:10:55+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26584273, 'toBlock': 26584282}
2025-09-18T20:10:55+0000 [EventActuator#debug] Scanning for blocks: 26584283 - 26584292, chunk size 9, last chunk scan took 0.05798840522766113, last logs found 0
2025-09-18T20:10:55+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26584283, 'toBlock': 26584292}
2025-09-18T20:10:55+0000 [operator#debug] 0x5d0105 performing phase 2 of DKG ritual #16 from blocktime 1758226140
```

✅ Encrypt/decrypt using ritual 16 with Alchemy free tier endpoint in three nodes

```
25-09-18T20:14:49+0000 [twisted.python.log#info] "<IP_REDACTED>" - - [18/Sep/2025:20:14:49 +0000] "GET /public_information HTTP/1.1" 200 755 "-" "python-requests/2.32.5"
2025-09-18T20:14:50+0000 [condition-eval#info] Evaluating access conditions {'version': '1.0.0', 'condition': {'chain': 80002, 'method': 'eth_getBalance', 'parameters': ['0x210eeAC07542F815ebB6FD6689637D8cA2689392', 'latest'], 'returnValueTest': {'comparator': '==', 'value': 0}, 'conditionType': 'rpc'}}
```

✅ One of the nodes with Alchemy free tier endpoint is stopped for a while. After that, it is started so Ritual Tracker has to catch up. It takes 31 seconds to catch up

```
2025-09-18T20:28:21+0000 [RitualTracker#debug] (0x131617) Scanning events in block range 26583067 - 26584867
2025-09-18T20:28:21+0000 [EventActuator#debug] Scanning for blocks: 26583067 - 26583127, chunk size 60, last chunk scan took 0, last logs found 0
2025-09-18T20:28:21+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237']], 'fromBlock': 26583067, 'toBlock': 26583127}
2025-09-18T20:28:22+0000 [EventActuator#warn] eth_getLogs API call failed for range 26583067 - 26583127 (60 blocks) on attempt 1/3: 400 Client Error: Bad Request for url: https://polygon-amoy.g.alchemy.com/v2/<redacted>
2025-09-18T20:28:22+0000 [EventActuator#warn] Alchemy free tier detected. Retrying with range 26583067 - 26583076 (9 blocks)
2025-09-18T20:28:22+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237']], 'fromBlock': 26583067, 'toBlock': 26583076}
2025-09-18T20:28:22+0000 [EventActuator#debug] Scanning for blocks: 26583077 - 26583086, chunk size 9, last chunk scan took 1.5555896759033203, last logs found 0
2025-09-18T20:28:22+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237']], 'fromBlock': 26583077, 'toBlock': 26583086}
2025-09-18T20:28:22+0000 [EventActuator#debug] Scanning for blocks: 26583087 - 26583096, chunk size 9, last chunk scan took 0.17902565002441406, last logs found 0
2025-09-18T20:28:22+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237']], 'fromBlock': 26583087, 'toBlock': 26583096}
2025-09-18T20:28:22+0000 [EventActuator#debug] Scanning for blocks: 26583097 - 26583106, chunk size 9, last chunk scan took 0.1728065013885498, last logs found 0
2025-09-18T20:28:22+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237']], 'fromBlock': 26583097, 'toBlock': 26583106}
2025-09-18T20:28:23+0000 [EventActuator#debug] Scanning for blocks: 26583107 - 26583116, chunk size 9, last chunk scan took 0.21060943603515625, last logs found 0
2025-09-18T20:28:23+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237']], 'fromBlock': 26583107, 'toBlock': 26583116}

...

2025-09-18T20:28:52+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237']], 'fromBlock': 26584867, 'toBlock': 26584867}
2025-09-18T20:28:52+0000 [RitualTracker#debug] Scanned total of 3 events, in 31.606645822525024 seconds, total 181 chunk scans performed
```

manumonti reviewed (approved):
I've reviewed the new changes and tests and it loooks great! 🙌

derekpierre committed:
Add acceptance test that runs using an actual alchemy polygon rpc endpoint that is under free tier.
The test only runs if the appropriate env variable is set - it is automatically set for CI.
